File: C/Algorithm/Algorithm I/21. Merge Two Sorted Lists.c
/*
 * Merges two sorted singly-linked lists into a single sorted list.
 *
 * The 'mergeTwoLists' function takes two pointers to the heads of two sorted
 * singly-linked lists, 'list1' and 'list2', as input. It merges the two lists into a
 * single sorted list in ascending order, using a new merged list that is constructed
 * iteratively. The function compares the values of nodes from 'list1' and 'list2', and
 * appends the smaller value node to the merged list. The merged list is built using the
 * 'mergedList' pointer, and the function returns a pointer to the head of the merged
 * list.
 *
 * Parameters:
 * - list1: A pointer to the head node of the first sorted singly-linked list.
 * - list2: A pointer to the head node of the second sorted singly-linked list.
 *
 * Returns:
 * A pointer to the head node of the merged sorted singly-linked list.
 *
 * This function efficiently merges the two sorted lists by iterating through both lists
 * and selectively appending nodes to the merged list. The final merged list contains
 * all elements from both input lists in sorted order.
 */
struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
    struct ListNode head;
    struct ListNode* mergedList = &head;

    while (list1 && list2) {
        if (list1->val < list2->val) {
            mergedList->next = list1;
            mergedList = mergedList->next;
            list1 = list1->next;
        }

        else {
            mergedList->next = list2;
            mergedList = mergedList->next;
            list2 = list2->next;
        }
    }

    if (list1) 
        mergedList->next = list1;
    
    else 
        mergedList->next = list2;

    return head.next;
}

File: Java/Algorithm/Algorithm I/21. Merge Two Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

  /**
   * Merges two sorted singly-linked lists into one sorted singly-linked list.
   *
   * @param list1 The head node of the first sorted linked list.
   * @param list2 The head node of the second sorted linked list.
   * @return The head node of the merged sorted linked list.
   */
  public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
    ListNode head = new ListNode();
    ListNode mergedList = head;

    while (list1 != null && list2 != null) {
      if (list1.val < list2.val) {
        mergedList.next = list1;
        list1 = list1.next;
      } 
      
      else {
        mergedList.next = list2;
        list2 = list2.next;
      }

      mergedList = mergedList.next;
    }

    if (list1 != null) mergedList.next = list1;
    if (list2 != null) mergedList.next = list2;

    return head.next;
  }
}
